- adds a first version of the upgrade guide for v3 (#624)
microsoftgraph/msgraph-sdk-java: docs/upgrade-to-v3.md

# Microsoft Graph Java SDK v3 changelog and upgrade guide

Welcome to the Microsoft Graph Java SDK migration guide to v3. The purpose of this document is to list out any breaking change and migration work SDK users might run into while upgrading to the latest version of the SDK.

## Breaking changes

> To improve the development experience of provided by the SDK, it was necessary to make the following breaking changes in addition to the changes listed in the migration guide below:

Reviewer: ```suggestion
To improve the development experience provided by the SDK, it was necessary to make the following breaking changes in addition to the changes listed in the migration guide below:
```

1. Remove any dependency on `microsoft-graph-auth` and replace them by one of the providers available in azure-identity.
1. Remove any dependency on `microsoft-graph-android-auth` and replace them by one of the providers available in azure-identity.
1. Reconsider the need for custom authentication providers (implementation of `IAuthenticationProvider` or `ICoreAuthenticationProvider`) that might be part of your project and use a provider available in azure-identity whenever possible.

### Single authentication provider interface

Due to historical reasons, the SDK provided two authentication provider interfaces `ICoreAuthenticationProvider` and `IAuthenticationProvider`. This was creating confusion and duplication and has been resolved to a single `IAuthenticationProvider` interface. If you implemented any of those interfaces you should:

1. Replace any reference to `ICoreAuthenticationProvider` by `IAuthenticationProvider`.
1. Update the implementation to match the new interface.

> 1. Replace any reference to `com.microsoft.graph.

Reviewer: ```suggestion
1. Replace any reference to `com.microsoft.graph`.
```

> > Note: if your project implements its own authentication provider, you might want to review the list of providers available with Azure identity to review whether a custom provider is still required. If a custom provider is still required, you should extend `BaseAuthenticationProvider` which provides a method to check whether a spefic request should be authenticated based on it's URL.

Reviewer: ```suggestion
> Note: if your project implements its own authentication provider, you might want to review the list of providers available with Azure identity to review whether a custom provider is still required. If a custom provider is still required, you should extend `BaseAuthenticationProvider` which provides a method to check whether a spefic request should be authenticated based on its URL.
```

             .worksheets("{id|name}")
             .range("A1:B2");
   ```

   Now becomes:

   ```Java
   final WorkbookRange workbookRange = graphClient.me()
             .drive()
             .items("{id}")
             .workbook()
             .worksheets("{id|name}")
             .range(WorkbookWorksheetRangeParameterSet
               .newBuilder()
               .withAddress("A1:B2")
               .build());
   ```

### Futures vs Callbacks

To provide a standard and modern way of executing requests asynchronuously to SDK users, the custom defined `ICallback` interface has been removed and the API now returns `CompletableFuture`.

1. Replace any of the code built the following way

    ```Java
    graphClient
    .me()
    .drive()
    .buildRequest()
    .get(new ICallback<Drive>() {
      @Override
      public void success(final Drive result) {
        System.out.println("Found Drive " + result.id);
      }
    });
    ```

    By:

    ```Java
    graphClient
    .me()
    .drive()
    .buildRequest()
    .getAsync()
    .thenApply(result -> {
        System.out.println("Found Drive " + result.id);
    });
    ```

### Non-generated code moved to the core library

> The service library (microsoft-graph) was designed to contain only code that has been generated automatically from the API metadata. All code that has been manually handcrafted should live in the core library (microsoft-graph-core). This principle makes reusability of code easier for consumers and it was not respected until this new version of the SDK. All the code living under anything but `com.microsoft.graph.models` or `.requests` have been moved to the same package name in the core asset. This move should be transparent to consummers expect for the following that needs to be updated:

Reviewer: We have enough context to deduce the full namespace, but it may not be that straightforward for someone occasionally using the SDK.
```suggestion
The service library (microsoft-graph) was designed to contain only code that has been generated automatically from the API metadata. All code that has been manually handcrafted should live in the core library (microsoft-graph-core). This principle makes reusability of code easier for consumers and it was not respected until this new version of the SDK. All the code living under anything but `com.microsoft.graph.models` or `com.microsoft.graph.requests` have been moved to the same package name in the core asset. This move should be transparent to consumers, except for the following that needs to be updated:
```

1. Replace any reference to `ChunkUploadProvider` from `com.microsoft.graph.requests.extensions` to `com.microsoft.graph.concurrency`.
1. Replace any reference to `CustomRequestBuilder`, `DateOnly`, `TimeOfDay`, and `Multipart` from `com.microsoft.graph.(models|requests).extensions` to `com.microsoft.graph.core`.

### Removal of connection configuration

As the connection configuration evolved over time, the connection configuration infrastructure had become redundant with the native library configuration capabilities:

1. Remove any reference to `IConnectionConfiguration` and re-implement the configuration using [Customize the Microsoft Graph SDK service client](https://docs.microsoft.com/graph/sdks/customize-client?tabs=java).

### Dependencies upgrade

- OkHttpClient has been upgraded from 3.X to 4.9.0.
- Google Guava has been upgraded from 20 to 29.

No action needs to be taken unless the project has a direct reference to an older version of those libraries, in which case the references need to be upgraded to match the versions listed here.

### float properties have been replaced by Float

> As the OData specification madates support for null values for numeric values, any property that previously was implemented using a float primitive type is now using a Float object type:

Reviewer: ```suggestion
As the OData specification mandates support for null values for numeric values, any property that previously was implemented using a float primitive type is now using a Float object type:
```

1. Replace any assignment from float to Float and check for null values.

### Merge of GraphServiceClient and BaseGraphServiceClient

To simplify the object model of the SDK, `BaseGraphServiceClient` has been merged into `GraphServiceClient`:

1. Replace any reference to `BaseGraphServiceClient` by `GraphServiceClient`.

### Generic refactoring

> Requests, request builders, and responses have been refactored to take advantage of generic types. This maximizes code reusability improving performances, binary size and more. These changes should be transparent to most SDK consumers but it's possible impact developers calling the SDK using reflection techniques.

Reviewer: ```suggestion
Requests, request builders, and responses have been refactored to take advantage of generic types. This maximizes code reusability improving performances, binary size and more. These changes should be transparent to most SDK consumers but it's possible to impact developers calling the SDK using reflection techniques.
```

    final Request batchRequest = new Request.Builder()
    .url("https://graph.microsoft.com/v1.0/$batch")
    .post(RequestBody.create(MediaType.parse("application/json"), batchContent))
    .build();
    final Response batchResponse = client.newCall(batchRequest).execute();
    final MSBatchResponseContent batchResponseContent = new MSBatchResponseContent(batchResponse);
    final Response userResponse = batchResponseContent.getResponseById(meGetId);
    ```

    By:

    ```Java
    final MSBatchRequestContent batchRequestContent = new MSBatchRequestContent();
    final String meGetId = batchRequestContent.addbatchRequestStep(graphClient.me().buildRequest().withHttpMethod(HttpMethod.GET).getHttpRequest());
    final MSBatchResponseContent batchResponseContent = batchRequestContent.execute(graphClient);
    final Response userResponse = batchResponseContent.getResponseById(meGetId);
    ```

> Note: this change also replaces the `MSBatchRequestContent` constuctor accepting a list of steps by a vararg constructor and `getArrayOfDependsOnIds` by `getDependsOnIds` (HashSet\<String>).

### IJsonBackedObject interface

> The interface has been simplified to remove the `getRawObject` and `getSerializer` methods. This avoid entity and collection objects storing a copy of the JSON on top the the properties they already have, which improves drastically the memory impacts:

Reviewer: ```suggestion
The interface has been simplified to remove the `getRawObject` and `getSerializer` methods. This avoid entity and collection objects storing a copy of the JSON on top of the properties they already have, which improves drastically the memory impacts:
```
